packages/@lwc/engine/src/framework/vm.ts

```diff
@@ -142,49 +137,43 @@ export function rerenderVM(vm: VM) {
     rehydrate(vm);
 }
 
-export function appendRootVM(vm: VM) {
-    runConnectedCallback(vm);
-    rehydrate(vm);
-}
-
 export function appendVM(vm: VM) {
+    runConnectedCallback(vm);
     rehydrate(vm);
 }
 
 // just in case the component comes back, with this we guarantee re-rendering it
 // while preventing any attempt to rehydration until after reinsertion.
-function resetComponentStateWhenRemoved(vm: VM) {
+function reset(vm: VM) {
     const { state } = vm;
     if (state !== VMState.disconnected) {
         const { oar, tro } = vm;
-        // Making sure that any observing record will not trigger the rehydrated on this vm
+        // Making sure that any observing record will not trigger the rehydration on this vm
         tro.reset();
         // Making sure that any observing accessor record will not trigger the setter to be reinvoked
         for (const key in oar) {
             oar[key].reset();
         }
-        runDisconnectedCallback(vm);
-        // Spec: https://dom.spec.whatwg.org/#concept-node-remove (step 14-15)
-        runShadowChildNodesDisconnectedCallback(vm);
-        runLightChildNodesDisconnectedCallback(vm);
+    }
+    if (isFalse(vm.isDirty)) {
+        // this guarantees that if the component is reused/reinserted, it will be re-rendered.
+        // it should be re-rendered because we are disconnecting the reactivity
+        // linking, so mutations are not automatically reflected on the state
+        // of disconnected components.
+        vm.isDirty = true;
     }
 }
 
-// this method is triggered by the diffing algo only when a vnode from the
-// old vnode.children is removed from the DOM.
+// this method is triggered by the removal of a element from the DOM.
 export function removeVM(vm: VM) {
     if (process.env.NODE_ENV !== 'production') {
         assert.isTrue(
             vm.state === VMState.connected || vm.state === VMState.disconnected,
             `${vm} must have been connected.`
         );
     }
-    resetComponentStateWhenRemoved(vm);
-}
-
-// this method is triggered by the removal of a root element from the DOM.
-export function removeRootVM(vm: VM) {
-    resetComponentStateWhenRemoved(vm);
+    reset(vm);
```

Comment:
I'm trying to remember why do we need to reset the template when disconnecting... I think we don't need anymore, we were doing that to be able to trigger the removal of the vm. I think it is safe to keep the template running.

Reply:
if I'm correct here, this might be able to give us the numbers we want... or even better.

Reply:
`reset` also guarantees that any changes to observed records during `disconnectedCallback` do not trigger a rehydration.

Reply:
ok ok, reset as implemented today is fine... probably rename it to `resetReactivity`, `resetReactiveListeners` or something like that is more specific. I think I was confused with `resetShadowRoot`, which is now gone.

Reply:
My point was that when removing things from the DOM, there is no need to remove anything from shadowRoot, or traverse it in any way, because the browser will effectively run the disconnect.